src/text_analysis_source_impl.rs

```diff
@@ -0,0 +1,203 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! A custom implementation of the "text analysis source" interface so that
+//! we can convey data to the `FontFallback::map_characters` method.
+
+#![allow(non_snake_case)]
+
+use std::borrow::Cow;
+use std::cell::UnsafeCell;
+use std::mem;
+use std::ptr::null;
+use std::sync::atomic::AtomicUsize;
+use winapi::ctypes::wchar_t;
+use winapi::shared::basetsd::UINT32;
+use winapi::shared::guiddef::REFIID;
+use winapi::shared::minwindef::{BOOL, FALSE, TRUE, ULONG};
+use winapi::shared::winerror::{E_INVALIDARG, S_OK};
+use winapi::um::dwrite::{DWRITE_NUMBER_SUBSTITUTION_METHOD, DWRITE_READING_DIRECTION,
+    IDWriteNumberSubstitution, IDWriteTextAnalysisSource, IDWriteTextAnalysisSourceVtbl,
+};
+use winapi::um::unknwnbase::{IUnknown, IUnknownVtbl};
+use winapi::um::winnt::HRESULT;
+
+use helpers::ToWide;
+use super::DWriteFactory;
+use com_helpers::{Com, UuidOfIUnknown};
+use comptr::ComPtr;
+
+/// The Rust side of a custom text analysis source implementation.
+pub trait TextAnalysisSourceImpl {
```

Comment:
nit: Kind of odd to name a trait `Impl`. Sometimes I use the word `Methods` for this, like `TextAnalysisSourceMethods`…

Reply:
Sure, I was pretty unsure about this.